File: kademlia/src/main/java/kademlia/KeyValueContent.java
package kademlia;

import com.google.gson.Gson;
import kademlia.dht.KadContent;
import kademlia.node.KademliaId;
import kademlia.simulations.DHTContentImpl;

import java.io.Serializable;

public class KeyValueContent implements KadContent, Serializable {
    private static final long serialVersionUID = 1L;
    private final String TYPE = "KeyValueContentImpl";
    private long updateTs = 0L;
    private String ownerId = "";
    private long createTs = 0L;
    private KademliaId key;
    private String data = "";

    public KeyValueContent(String _key, String _value, String _owner) {
        createTs = System.currentTimeMillis() / 1000L;
        key = new KademliaId(_key);
        ownerId = _owner;
        data = _value;
    }

    public void setUpdated() {
        updateTs = System.currentTimeMillis() / 1000L;
    }

    public String getData() {
        return data;
    }

    @Override
    public KademliaId getKey() {
        return key;
    }

    @Override
    public String getType() {
        return DHTContentImpl.TYPE;
    }

    @Override
    public long getCreatedTimestamp() {
        return createTs;
    }

    @Override
    public long getLastUpdatedTimestamp() {
        return updateTs;
    }

    @Override
    public String getOwnerId() {
        return ownerId;
    }

    @Override
    public byte[] toSerializedForm() {
        Gson gson = new Gson();
        return gson.toJson(this).getBytes();
    }

    @Override
    public KadContent fromSerializedForm(byte[] p) {
        Gson gson = new Gson();
        return gson.fromJson(new String(p), DHTContentImpl.class);
    }
}

File: kademlia/src/main/java/kademlia/KeyValueContent.kt
package kademlia

import java.nio.charset.StandardCharsets
import com.google.gson.Gson
import kademlia.dht.KadContent
import kademlia.node.KademliaId
import kademlia.simulations.DHTContentImpl


class KeyValueContent(_key: String, _value: String, _owner: String): KadContent {
    @Transient
    val TYPE = "KeyValueContentImpl"

    private var updateTs = 0L
    private var ownerId = ""
    private var createTs = 0L
    private var key: KademliaId
    private var data: String = ""

    init{
        createTs = System.currentTimeMillis() / 1000L
        key = KademliaId(_key)
        ownerId = _owner
        data = _value
    }

    fun setUpdated() {
        updateTs = System.currentTimeMillis() / 1000L
    }

    fun getData():String {
        return data
    }
    override fun getKey(): KademliaId {
        return key
    }

    override fun getType(): String {
        return DHTContentImpl.TYPE
    }

    override fun getCreatedTimestamp(): Long {
        return createTs
    }

    override fun getLastUpdatedTimestamp(): Long {
        return updateTs
    }

    override fun getOwnerId(): String {
        return ownerId
    }

    override fun toSerializedForm(): ByteArray {
        val gson = Gson()
        return gson.toJson(this).toByteArray()

    }

    override fun fromSerializedForm(p: ByteArray?): KadContent {
        val gson = Gson()
        return gson.fromJson(String(p!!), DHTContentImpl::class.java)
    }
}